### PR TITLE
LWDEV-8243 Use context facility in CashInCashOut/CashTransferService service

### PR DESCRIPTION
**Pull request by rocketmannot**

### CODE REVIEW

File: src/test/kotlin/com/lykke/matching/engine/services/CashOperationServiceTest.kt [hunk 1]
```diff
@@ -198,22 +204,23 @@ class CashOperationServiceTest: AbstractTest() {
         assertEquals("-24.00", operation.reservedVolume)
         assertEquals("Asset1", operation.asset)
 
-        reservedCashInOutOperationService.processMessage(buildReservedCashInOutWrapper("Client3", "Asset1", -30.0))
-        reservedBalance = testWalletDatabaseAccessor.getReservedBalance("Client3", "Asset1")
+        val messageWrapper1 = buildReservedCashInOutWrapper("Client3", "Asset1", -30.0)
+        reservedCashInOutOperationService.processMessage(messageWrapper1)
         assertEquals(BigDecimal.valueOf(26.0), reservedBalance)
```
Comment (barsukovskii) [inline src/test/kotlin/com/lykke/matching/engine/services/CashOperationServiceTest.kt]: read balance again before assert

File: src/test/kotlin/com/lykke/matching/engine/services/validator/business/CashTransferOperationBusinessValidatorTest.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        }
+    }
+
+    @Test(expected = ValidationException::class)
+    fun testLowBalance() {
+        //given
+        testBalanceHolderWrapper.updateBalance(CLIENT_NAME1, ASSET_ID, 100.0)
+        testBalanceHolderWrapper.updateReservedBalance(CLIENT_NAME1, ASSET_ID, 50.0)
+        val cashTransferOperationBuilder = getCashTransferOperationBuilder()
+        cashTransferOperationBuilder.overdraftLimit = -70.0
```
Comment (barsukovskii) [inline src/test/kotlin/com/lykke/matching/engine/services/validator/business/CashTransferOperationBusinessValidatorTest.kt]: Strange test of transfer with volume=0 and negative overdraftLimit. Let's check transfer with volume=60.0 (for example) and overdraft=0 (or 5.0). Perhaps we should validate also overdraftLimit value in preprocessor

File: src/main/kotlin/com/lykke/matching/engine/daos/WalletOperation.kt [hunk 3]
```diff
@@ -12,7 +13,8 @@ data class WalletOperation(
         val dateTime: Date,
         val amount: BigDecimal,
         val reservedAmount: BigDecimal = BigDecimal.ZERO,
-        val isFee: Boolean = false
+        val isFee: Boolean = false,
+        val feeInstructions: List<NewFeeInstruction>? = null
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/daos/WalletOperation.kt]: fixed

File: src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashInOutContextParser.kt [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        val message = ProtocolMessages.CashInOutOperation.parseFrom(messageWrapper.byteArray)
+
+        messageWrapper.id = message.id
+        messageWrapper.messageId = message.messageId
+
+        messageWrapper.context = CashInOutContext(message.id,
+                if (message.hasMessageId()) message.messageId else message.id,
+                operationId,
+                message.clientId,
+                ProcessedMessage(MessageType.CASH_IN_OUT_OPERATION.type, message.timestamp, message.messageId),
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashInOutContextParser.kt]: fixed

File: src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashTransferContextParser.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        messageWrapper.id = message.id
+        messageWrapper.messageId = message.messageId
+
+        messageWrapper.context =
+                CashTransferContext(
+                        if (message.hasMessageId()) message.messageId else message.id,
+                        transferOperation,
+                        assetsHolder.getAssetAllowNulls(transferOperation.asset),
+                        transferOperation.asset,
+                        ProcessedMessage(MessageType.CASH_TRANSFER_OPERATION.type, message.timestamp, message.messageId),
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashTransferContextParser.kt]: fixed

File: src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashTransferContextParser.kt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        val feeInstructions = NewFeeInstruction.create(message.feesList)
+
+        val transferOperation = TransferOperation(UUID.randomUUID().toString(), message.id,
+                message.fromClientId, message.toClientId,
+                message.assetId, Date(message.timestamp),
+                BigDecimal.valueOf(message.volume), BigDecimal.valueOf(message.overdraftLimit),
+                listOfFee(feeInstruction, feeInstructions))
+
+        messageWrapper.id = message.id
+        messageWrapper.messageId = message.messageId
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashTransferContextParser.kt]: fixed

File: src/test/kotlin/com/lykke/matching/engine/services/CashOperationServiceTest.kt (hunk 1 above)
Comment (rocketmannot) [reply src/test/kotlin/com/lykke/matching/engine/services/CashOperationServiceTest.kt]: fixed

File: src/test/kotlin/com/lykke/matching/engine/services/validator/business/CashTransferOperationBusinessValidatorTest.kt (hunk 2 above)
Comment (rocketmannot) [reply src/test/kotlin/com/lykke/matching/engine/services/validator/business/CashTransferOperationBusinessValidatorTest.kt]: fixed, new validation overdraft limit is not negative is added

File: src/test/kotlin/com/lykke/matching/engine/services/CashOperationServiceTest.kt [hunk 7]
```diff
@@ -198,22 +204,24 @@ class CashOperationServiceTest: AbstractTest() {
         assertEquals("-24.00", operation.reservedVolume)
         assertEquals("Asset1", operation.asset)
 
-        reservedCashInOutOperationService.processMessage(buildReservedCashInOutWrapper("Client3", "Asset1", -30.0))
+        val messageWrapper1 = buildReservedCashInOutWrapper("Client3", "Asset1", -30.0)
         reservedBalance = testWalletDatabaseAccessor.getReservedBalance("Client3", "Asset1")
+        reservedCashInOutOperationService.processMessage(messageWrapper1)
```
Comment (barsukovskii) [inline src/test/kotlin/com/lykke/matching/engine/services/CashOperationServiceTest.kt]: read reserveBalance (previous line) after this line

File: src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashInOutContextParser.kt [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+@Component
+class CashInOutContextParser(private val assetsHolder: AssetsHolder) : ContextParser<CashInOutParsedData> {
+    override fun parse(messageWrapper: MessageWrapper): CashInOutParsedData {
+        val operationId = UUID.randomUUID().toString()
+
+        val message = ProtocolMessages.CashInOutOperation.parseFrom(messageWrapper.byteArray)
+
+        messageWrapper.id = message.id
+        messageWrapper.messageId = if (message.hasMessageId()) message.messageId else message.id
+
```
Comment (barsukovskii) [inline src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashInOutContextParser.kt]: add initialization of messageWrapper.timestamp

File: src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashTransferContextParser.kt [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+        val transferOperation = TransferOperation(UUID.randomUUID().toString(), message.id,
+                message.fromClientId, message.toClientId,
+                message.assetId, Date(message.timestamp),
+                BigDecimal.valueOf(message.volume), BigDecimal.valueOf(message.overdraftLimit),
+                listOfFee(feeInstruction, feeInstructions))
+
+        messageWrapper.id = message.id
+        messageWrapper.messageId = if (message.hasMessageId()) message.messageId else message.id
+
```
Comment (barsukovskii) [inline src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashTransferContextParser.kt]: add initialization of messageWrapper.timestamp

File: src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashInOutContextParser.kt (hunk 8 above)
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashInOutContextParser.kt]: fixed

File: src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashTransferContextParser.kt (hunk 9 above)
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/incoming/parsers/impl/CashTransferContextParser.kt]: fixed

**Pull request merged**